### PR TITLE
msg server function and tests for MsgScheduleIBCClientUpgrade

### PR DESCRIPTION
**Pull request by charleenfei**
Please first review #4436 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/keeper/msg_server_test.go [hunk 1]
```diff
@@ -827,7 +831,74 @@ func (suite *KeeperTestSuite) TestUpdateClientParams() {
...
-func (*KeeperTestSuite) TestIBCSoftwareUpgrade() {
+func (suite *KeeperTestSuite) TestIBCSoftwareUpgrade() {
+	var (
+		expError error
+		msg      *clienttypes.MsgIBCSoftwareUpgrade
+	)
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
```
Comment (damiannolan) [inline modules/core/keeper/msg_server_test.go:842]: same as other PR, we can refactor this to be `expError` and use `expPass := expError == nil` in the test body! 👍

File: modules/core/keeper/msg_server_test.go [hunk 2]
```diff
@@ -827,7 +831,74 @@ func (suite *KeeperTestSuite) TestUpdateClientParams() {
...
+				ibctm.NewClientState(suite.chainA.ChainID, ibctesting.DefaultTrustLevel, ibctesting.TrustingPeriod, ibctesting.UnbondingPeriod, ibctesting.MaxClockDrift, clienttypes.NewHeight(1, 10), commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath),
+			)
+
+			suite.Require().NoError(err)
+
+			tc.malleate()
+
+			_, err = keeper.Keeper.IBCSoftwareUpgrade(*suite.chainA.App.GetIBCKeeper(), suite.chainA.GetContext(), msg)
+			if tc.expPass {
+				suite.Require().NoError(err)
```
Comment (damiannolan) [inline modules/core/keeper/msg_server_test.go:896]: Inside `expPass` we can also check that there exists an upgrade plan in the `x/upgrade` store and upgraded client state in the `upgradePath`

**Review by damiannolan — APPROVED** [review]
Looks super clean! Just left two suggestions on testing 👍

File: modules/core/02-client/types/msgs_test.go [hunk 3]
```diff
@@ -849,3 +851,74 @@ func (suite *TypesTestSuite) TestMsgIBCSoftwareUpgrade_GetSigners() {
 		}
 	}
 }
+
+// TestMsgIBCSoftwareUpgrade_ValidateBasic tests ValidateBasic for MsgIBCSoftwareUpgrade
+func (suite *TypesTestSuite) TestMsgIBCSoftwareUpgrade_ValidateBasic() {
```
Comment (charleenfei) [inline modules/core/02-client/types/msgs_test.go:856]: this test got removed in the last pr merge process.

File: modules/core/keeper/msg_server_test.go [hunk 4]
```diff
@@ -827,7 +831,86 @@ func (suite *KeeperTestSuite) TestUpdateClientParams() {
 }
 
 // TestIBCSoftwareUpgrade tests the IBCSoftwareUpgrade rpc handler
-func (*KeeperTestSuite) TestIBCSoftwareUpgrade() {
+func (suite *KeeperTestSuite) TestIBCSoftwareUpgrade() {
+	var (
+		expError error
```
Comment (colin-axner) [inline modules/core/keeper/msg_server_test.go:836]: super nit: I like defining the expError on the testCases, it makes it more explicit what we expect to occur and avoids potential mishaps of not setting an expected error

File: modules/core/keeper/msg_server_test.go [hunk 5]
```diff
@@ -827,7 +831,86 @@ func (suite *KeeperTestSuite) TestUpdateClientParams() {
...
+			path := ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.SetupClients(path)
+			validAuthority := suite.chainA.App.GetIBCKeeper().GetAuthority()
+			plan := upgradetypes.Plan{
+				Name:   "upgrade IBC clients",
+				Height: 1000,
+			}
+			// update trusting period
+			clientState := ibctm.NewClientState(suite.chainA.ChainID, ibctesting.DefaultTrustLevel, ibctesting.TrustingPeriod+100, ibctesting.UnbondingPeriod, ibctesting.MaxClockDrift, clienttypes.NewHeight(1, 10), commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath)
+
```
Comment (colin-axner) [inline modules/core/keeper/msg_server_test.go:882]: could also likely do:
```go
clientState := path.EndpointB.GetClientState()
clientState.TrustingPeriod+=100
```

**Review by colin-axner — APPROVED** [review]
LGTM! Left a tiny nit and a suggestion (implement if you wish!)

**Pull request merged**